amc/egl: smooth rendering of video frames

aslobodeniuk opened the pull request:
This code tries not to loose VSYNC on android.
To make it possible we follow next ideas:
- call releaseOutputBuffer immediatelly, so it
won't be dropped by SurfaceFlinger. Drops were
really happening before.
- set presentation timestamp for surface we're
going to render before eglSwapBuffers.
- timestamp is calculated to moment at which buffer
should be rendered in system time.
- sink must wake up and "enqueue" surface with
eglSwapBuffers at least ~20ms before the moment
at which buffer should be rendered

NOTE: this solution plays smooth only if
application uses SurfaceView.

aslobodeniuk committed:
amc/egl: smooth rendering of video frames

This code tries not to loose VSYNC on android.
To make it possible we follow next ideas:
- call releaseOutputBuffer immediatelly, so it
won't be dropped by SurfaceFlinger. Drops were
really happening before.
- set presentation timestamp for surface we're
going to render before eglSwapBuffers.
- timestamp is calculated to moment at which buffer
should be rendered in system time.
- sink must wake up and "enqueue" surface with
eglSwapBuffers at least ~20ms before the moment
at which buffer should be rendered

NOTE: this solution plays smooth only if
application uses SurfaceView.

jmcatena commented on ext/eglgles/gstegladaptation_egl.c:
```diff
@@ -439,8 +439,20 @@ gst_egl_adaptation_bind_API (GstEglAdaptationContext * ctx)
 {
+#if HAVE_ANDROID
+  /* display at next ms */
+  if (G_LIKELY (ts >= 0)) {
```
Please use GST_CLOCK_TIME_IS_VALID or GST_CLOCK_TIME_NONE to avoid issues with unsigned or signed values. We should use GstClockTime instead of gint64, and in that case this code will fail.

jmcatena commented on ext/eglgles/gsteglglessink.c:
```diff
@@ -1389,6 +1423,11 @@ gst_eglglessink_configure_caps (GstEglGlesSink * eglglessink, GstCaps * caps)
 
+#if HAVE_ANDROID
+  /* !!! FIXME: calculate based on frame rate. */
+  gst_base_sink_set_ts_offset (GST_BASE_SINK (eglglessink), -50000000l);
```
What this ts offset mean? We want to pass the buffer to eglgles before its ts (latency?), we do not want to change the presentation time. We have to ensure that video / audio sync is not affected!

jmcatena commented on gst-libs/gst/androidjni/gstjniamcdirectbuffer.c:
```diff
@@ -68,10 +70,25 @@ gst_jni_amc_direct_buffer_render (GstJniAmcDirectBuffer * buf)
     ret = gst_jni_call_void_method (env, buf->media_codec,
-        buf->release_output_buffer, buf->idx, TRUE);
+        buf->release_output_buffer_ts, buf->idx,
+        g_get_monotonic_time () * 1000);
```
This does not look like what we need.
We do not want to set the presentation time to now. It should be the presentation time output by the decoder (the buffer timestamp) adjusted for the system time base. And we should check that (adjusted_ts - now < 1s).

aslobodeniuk replied:
It's not the line that is swapping the buffer to be shown, but the line which is telling to put the image from the decoded buffer to decoder's surface as soon as possible. After that in a sink this we'll take a texture from this surface, and copy it to the window's surface with scaling and drawing the black borders.

jmcatena replied:
TY, I knew what releaseOutputBuffer in Android is. If we use the overload that takes a timestamp, like we are doing after the change, that is the time at which the amc sink will render to output, as long as the value is within 1s ahead of system time.
So I am very surprised to know that this is working.
Anyway, you stressed that these changes improved the behavior, and I approved as a temporary solution. We should keep investigating for a more correct solution afterwards, as this does not look correct.

ylatuya replied:
I think this timestamp has no effect at all in when it's decoded, this timestamp is the one used by SurfaceFlinger's buffer queue to decide when it has to be rendered. Also, if we are already changing the timestamp in gst_egl_adaptation_swap_buffers this will have no effect as we are already overriding it in the sink.
If we really want to set a timestamp here, this should never be the current system time, but the system time at which we want to buffer to be rendered, the same we end up setting in the sink.

aslobodeniuk replied:
In this implementation (egl) we have 2 surfaces: one to which we do swapBuffers and other, to which we do releaseOutputBuffer. We put "now" timestamp to releaseOutputBuffer to say we want the surface to be updated as soon as possible. It's not that surface user will see, but the other one, immediate, from which we are copying the image.

jmcatena reviewed (approved):
I do not think these changes are good enough.
But this has been found to solve an urgent customer issue and I'll approve as a temporary solution.
More work will be necessary to find and verify a better solution.

aslobodeniuk committed:
eglglessink: allow user to provide window late

Now if we are going to render and no window is
provided still, we're just going to complain in
the logs, but not refusing the work at all.
So if user wasn't lucky at some moment, even if
it was a user's fault, we can recover.

aslobodeniuk committed:
amcaudiodec: clean log garbage
